src/Mod/Sketcher/Gui/EditModeCoinManager.h

```diff
@@ -0,0 +1,293 @@
+ *
+ * EditModeCoinManager is designed to define the span of time in which ViewProviderSketch is in edit mode.
+ *
+ * In addition to the scenograph, EditModeCoinManager is responsible for keeping any necessary mapping between
+ * indices used at ViewProviderSketch level, and internal indexing used by EditModeCoinManager and its subclasses.
+ */
+class SketcherGuiExport EditModeCoinManager
+{
+    /** @brief      Class for monitoring changes in parameters affecting drawing and coin node generation
+    *  @details
+    *
+    * This nested class is a helper responsible for attaching to the parameters relevant for
+    * EditModeCoinManager and its helpers, initialising the EditModeCoinManager to the current configuration
+    * and handle in real time any change to their values.
+    */
+    class ParameterObserver : public ParameterGrp::ObserverType
+    {
+    private:
+        enum class OverlayVisibilityParameter {
+            BSplineDegree,
+            BSplineControlPolygonVisible,
+            BSplineCombVisible,
+            BSplineKnotMultiplicityVisible,
+            BSplinePoleWeightVisible
+        };
+
+    public:
+        ParameterObserver(EditModeCoinManager & client);
+        ~ParameterObserver();
+
+        void subscribeToParameters();
+
+        void unsubscribeToParameters();
+
+        /** Observer for parameter group. */
+        void OnChange(Base::Subject<const char*> &rCaller, const char * sReason) override;
+
+    private:
+        void initParameters();
+        void updateCurvedEdgeCountSegmentsParameter(const std::string & parametername);
+        void updateLineRenderingOrderParameters(const std::string & parametername);
+        void updateConstraintPresentationParameters(const std::string & parametername);
+        void updateElementSizeParameters(const std::string & parametername);
+        void updateColor(SbColor &sbcolor, const std::string &parametername);
+
+        template<OverlayVisibilityParameter visibilityparameter>
+        void updateOverlayVisibilityParameter(const std::string & parametername);
+
+    private:
+        std::map<std::string, std::function<void(const std::string &)>> str2updatefunction;
+        EditModeCoinManager &Client;
+    };
+
+public:
+    /** @brief This struct defines the information provided to other classes about preselection.
+     *
+     * @details
+     *
+     * PointIndex: Only Positive values corresponding to VertexId (are positive for both normal and external geometry)
+     * GeoIndex: Same values as GeoId of GeoElementId, except for axes which are not included. -1 represents an invalid curve.
+     *
+     * In other words valid values are 0,1,2,... for normal geometry and -3,-4,-5,... for external geometry
+     *
+     * Cross: Axes and RootPoint values as defined in the enum class.
+     *
+     */
+    struct PreselectionResult {
+        enum SpecialValues {
+            InvalidPoint = -1,
+            InvalidCurve = -1,

```

Comment:
Suspect value, -1 is used twice in an enum

Reply:
It is valid and intended. 

However, this is something I want to refactor further at a second stage. I have "improved" it, so that it is clear what the magic numbers are (and what the intention is). But, I still do not like the fact that preselection goes one way (separate cross, different values), selection another way, and this lack of unity makes every struct has its "Special Values".